packages/libp2p/test/core/random-walk.spec.ts

```diff
@@ -0,0 +1,169 @@
+import { stop } from '@libp2p/interface'
+import { defaultLogger } from '@libp2p/logger'
+import { createEd25519PeerId } from '@libp2p/peer-id-factory'
+import { multiaddr } from '@multiformats/multiaddr'
+import { expect } from 'aegir/chai'
+import delay from 'delay'
+import all from 'it-all'
+import drain from 'it-drain'
+import map from 'it-map'
+import take from 'it-take'
+import { stubInterface, type StubbedInstance } from 'sinon-ts'
+import { RandomWalk as RandomWalkClass } from '../../src/random-walk.js'
+import type { PeerRouting, PeerInfo, AbortOptions } from '@libp2p/interface'
+import type { RandomWalk } from '@libp2p/interface-internal'
+
+let port = 1234
+
+async function createRandomPeerInfo (): Promise<PeerInfo> {
+  port++
+
+  return {
+    id: await createEd25519PeerId(),
+    multiaddrs: [
+      multiaddr(`/ip4/123.123.123.123/tcp/${port}`)
+    ]
+  }
+}
+
+// eslint-disable-next-line require-yield
+async function * slowIterator (): any {
+  await delay(1000)
+}
+
+describe('random-walk', () => {
+  let randomwalk: RandomWalk
+  let peerRouting: StubbedInstance<PeerRouting>
+
+  beforeEach(async () => {
+    peerRouting = stubInterface<PeerRouting>()
+
+    randomwalk = new RandomWalkClass({
+      peerRouting,
+      logger: defaultLogger()
+    })
+  })
+
+  afterEach(async () => {
+    await stop(randomwalk)
+  })
+
+  it('should perform a random walk', async () => {
+    const randomPeer = await createRandomPeerInfo()
+
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * () {
+        yield randomPeer
+      }())
+      .onSecondCall().returns(slowIterator())
+
+    const peers = await all(take(randomwalk.walk(), 1))
+
+    expect(peers.map(peer => peer.id.toString())).to.include(randomPeer.id.toString())
+  })
+
+  it('should break out of a random walk', async () => {
+    let yielded = 0
+
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * (key, options?: AbortOptions) {
+        for (let i = 0; i < 100; i++) {
+          yielded++
+          yield await createRandomPeerInfo()
+          options?.signal?.throwIfAborted()
+        }
+      }())
+      .onSecondCall().returns(slowIterator())
+
+    await drain(take(randomwalk.walk(), 1))
+
+    expect(yielded).to.equal(1)
+  })
+
+  it('should keep walking until done', async () => {
+    const randomPeer1 = await createRandomPeerInfo()
+    const randomPeer2 = await createRandomPeerInfo()
+
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * () {
+        yield randomPeer1
+      }())
+      .onSecondCall().returns(async function * () {
+        yield randomPeer2
+      }())
+      .onThirdCall().returns(slowIterator())
```

Comment:
Ordinarily yes, though more like:

```
peerRouting.getClosestPeers
  .callsFake(async function * () {
    yield randomPeer1
      yield randomPeer2
  })
```

..but not in this test because here we want to assert that the query has been run twice, so we stub multiple invocations.